File: utilities/xpl-actions.py
#!/usr/bin/python3
import argparse
import sys
import signal
import os
import time
sys.path.append(sys.path[0]+'/../xPL-base')
import common

# ------------------------------------------------------------------------------
# constants
#
VENDOR_ID = 'dspc';             # from xplproject.org
DEVICE_ID = 'actions';          # max 8 chars
CLASS_ID = 'actions';           # max 8 chars

LOG_FILE_LENGTH = 100

INDENT = '  '
SEPARATOR = 80 * '-'

# ------------------------------------------------------------------------------
# command line arguments
#
parser = argparse.ArgumentParser()
                                                                     # verbosity
parser.add_argument(
    '-v', '--verbose', action='store_true', dest='verbose',
    help = 'verbose console output'
)
                                                                 # Ethernet port
parser.add_argument(
    '-p', '--port', default=50000,
    help = 'the clients base UDP port'
)
                                                                   # instance id
parser.add_argument(
    '-n', '--id', default=common.xpl_build_automatic_instance_id(),
    help = 'the instance id (max. 16 chars)'
)
                                                               # heartbeat timer
parser.add_argument(
    '-t', '--timer', default=5,
    help = 'the heartbeat interval in minutes'
)
                                                                 # startup delay
parser.add_argument(
    '-w', '--wait', default=0,
    help = 'the startup sleep interval in seconds'
)
                                                             # scripts directory
parser.add_argument(
    '-d', '--directory',
    default=os.path.dirname(os.path.realpath(__file__))+os.sep+'actions',
    help = 'the directory containing the action scripts'
)
                                                             # scripts directory
parser.add_argument(
    '-l', '--logFile', default='/tmp/xpl-actions.log',
    help = 'the directory containing the action scripts'
)
                                                  # parse command line arguments
parser_arguments = parser.parse_args()
verbose = parser_arguments.verbose
Ethernet_base_port = parser_arguments.port
instance_id = parser_arguments.id
heartbeat_interval = int(parser_arguments.timer)
startup_delay = int(parser_arguments.wait)
actions_directory = parser_arguments.directory
log_file_spec = parser_arguments.logFile

debug = False

# ==============================================================================
# Internal functions
#

#-------------------------------------------------------------------------------
# Execute a command together with its arguments
#
def execute_command(command, body) :
                                                # get arguments from xPL message
    arguments = '';
    for key in body :
        if key != 'command' :
            if len(key) == 1 :
                arguments = arguments + " -%s %s" %(key, body[key])
            else :
                arguments = arguments + " --%s %s" %(key, body[key])
                                                             # check for command
    action = os.sep.join([actions_directory, command])
    if os.path.isfile(action) :
        full_command = command + arguments
        full_action = action + arguments
        if verbose :
            print(INDENT + "Received command: \"%s\"\n" % full_command);
                                                         # limit log file length
        if os.path.isfile(log_file_spec) :
            log_file_lines = open(log_file_spec, "r").read().split("\n")
            log_file_lines.extend(['', "%s :" % full_command, '', ''])
            open(log_file_spec, "w").write(
                "\n".join(log_file_lines[-LOG_FILE_LENGTH:])
            )
                                                               # execute command
        os.system(full_action + ' >> ' + log_file_spec)

# ------------------------------------------------------------------------------
# catch ctrl-C interrupt
#
end = False

def ctrl_C_handler(sig, frame):
    global end
    end = True
    print('')

signal.signal(signal.SIGINT, ctrl_C_handler)

# ==============================================================================
# main script
#
                                                                 # startup delay
time.sleep(startup_delay);
                                                                # xPL parameters
xpl_id = common.xpl_build_id(VENDOR_ID, DEVICE_ID, instance_id);
xpl_ip = common.xpl_find_ip()
                                                             # create xPL socket
(client_port, xpl_socket) = common.xpl_open_socket(
    common.XPL_PORT, Ethernet_base_port
)
                                                    # display working parameters
if verbose :
    os.system('clear||cls')
    print(SEPARATOR)
    print("Ready to launch commands based on xPL messages")
    print(INDENT + "class id    : %s" % CLASS_ID)
    print(INDENT + "instance id : %s" % instance_id)
    print(INDENT + "actions directory: \"%s\"\n" % actions_directory);
    print()

# ..............................................................................
                                                                  # main loop
timeout = 1;
last_heartbeat_time = 0;
last_message_time = 0;

while not end :
                                                 # check time and send heartbeat
    last_heartbeat_time = common.xpl_send_heartbeat(
        xpl_socket, xpl_id, xpl_ip, client_port,
        heartbeat_interval, last_heartbeat_time
    )
                                              # get xpl-UDP message with timeout
    (xpl_message, source_address) = common.xpl_get_message(xpl_socket, timeout);
                                                           # process XPL message
    if (xpl_message) :
        (xpl_type, source, target, schema, body) = \
            common.xpl_get_message_elements(xpl_message)
        if schema == CLASS_ID + '.basic' :
            if xpl_type == 'xpl-cmnd' :
                if common.xpl_is_for_me(xpl_id, target) :
                    if verbose :
                        print("Received command from %s" % source)
                    command = body['command'];
                    if command :
                        execute_command(command, body)
                                                             # delete xPL socket
common.xpl_disconnect(xpl_socket, xpl_id, xpl_ip, client_port)

File: utilities/xpl-actions.pl
#!/usr/bin/perl

use File::Basename;
use FindBin;                            # find the script's directory
use lib "$FindBin::Bin/../xPL-base";    # add path for common lib
use common;


################################################################################
# constants
#
$vendor_id = 'dspc';            # from xplproject.org
$device_id = 'actions';         # max 8 chars
$class_id = 'actions';          # max 8 chars

$separator = '-' x 80;
$separator2 = '=' x 80;
$indent = ' ' x 2;

#-------------------------------------------------------------------------------
# global variables
#
my %configuration;
$configuration{'baseDirectory'} = dirname($0) . '/actions';


################################################################################
# Input arguments
#
use Getopt::Std;
my %opts;
getopts('hvp:n:t:w:d:', \%opts);

die("\n".
    "Usage: $0 [options]\n".
    "\n".
    "Options:\n".
    "${indent}-h      display this help message\n".
    "${indent}-v      verbose\n".
    "${indent}-p port the base UDP port\n".
    "${indent}-n id   the instance id (max. 16 chars)\n".
    "${indent}-t mins the heartbeat interval in minutes\n".
    "${indent}-w secs the startup sleep interval\n".
    "${indent}-d dir  the directory containing the action scripts\n".
    "\n".
    "Launches scripts based on xPL messages.\n".
    "\n".
    "More information with: perldoc $0\n".
    "\n".
    ""
   ) if ($opts{h});
my $verbose = $opts{v};
my $client_base_port = $opts{'p'} || 50000;
my $startup_sleep_time = $opts{'w'} || 0;

my $instance_id = $opts{'n'} || xpl_build_automatic_instance_id;
my $heartbeat_interval = $opts{'t'} || 5;

$configuration{'baseDirectory'} = $opts{'d'} || $configuration{'baseDirectory'};


################################################################################
# Internal functions
#

#-------------------------------------------------------------------------------
# Execute a command together with its arguments
#
sub execute_command {

  my ($base_directory, $verbose, %body) = @_;
                                                  # get command from xPL message
  my $action = $body{'command'};
                                                # get arguments from xPL message
  my $arguments = '';
  foreach $name (keys(%body)) {
    if ($body{$name} =~ m/\s/) {
      $body{$name} = "'$body{$name}'";
    }
    if ($name ne 'command') {
      if (length($name) == 1) {
        $arguments .= "-$name $body{$name} "
      } else {
        $arguments .= "--$name $body{$name} "
      }
    }
  }
  $arguments =~ s/\s\Z//;
                                                             # check for command
  if ($verbose > 0) {
    print $indent, "Received command: \"$action $arguments\"\n";
  }
  my $command_exists = 0;
  $action = "$base_directory/$action";
  if (-e $action) {
    $action = "$action $arguments";
    $command_exists = 1;
    system("$action >/dev/null 2>&1 &");
  }
                                                                  # display info
  if ($verbose > 0) {
    if ($command_exists == 0) {
      print $indent x 2, "command not found\n";
    }
    else {
      print $indent x 2, "executed $action\n";
    }
    print "\n";
  }
}


################################################################################
# Catch control-C interrupt
#
$SIG{INT} = sub{ $xpl_end++ };

################################################################################
# Main script
#
sleep($startup_sleep_time);
                                                                # xPL parameters

my $xpl_id = xpl_build_id($vendor_id, $device_id, $instance_id);
my $xpl_ip = xpl_find_ip;
                                                             # create xPL socket
my ($client_port, $xpl_socket) = xpl_open_socket($xpl_port, $client_base_port);
                                                    # display working parameters
if ($verbose > 0) {
  system("clear");
  print("$separator\n");
  print "Ready to launch commands based on xPL messages.\n";
  print($indent . "class id: $class_id\n");
  print($indent . "xPL id: $xpl_id\n");
  print($indent . "actions directory: \"$configuration{'baseDirectory'}\"\n");
	print("\n");
}

#-------------------------------------------------------------------------------
# Main loop
#
my $timeout = 1;
my $last_heartbeat_time = 0;

while ( (defined($xpl_socket)) && ($xpl_end == 0) ) {
                                                 # check time and send heartbeat
  $last_heartbeat_time = xpl_send_heartbeat(
    $xpl_socket, $xpl_id, $xpl_ip, $client_port,
    $heartbeat_interval, $last_heartbeat_time
  );
                                              # get xpl-UDP message with timeout
  my ($xpl_message) = xpl_get_message($xpl_socket, $timeout);
                                                           # process XPL message
  if ($xpl_message) {
    my ($type, $source, $target, $schema, %body) = xpl_get_message_elements(
      $xpl_message
    );
    if ($schema eq "$class_id.basic") {
      if ( xpl_is_for_me($xpl_id, $target) ) {
        if ($type eq 'xpl-cmnd') {
                                                               # process command
          if ($verbose > 0) {
            print("\n");
            print("Received command from \"$source\"\n");
          }
          my $command = $body{'command'};
          if ($command) {
            if ($verbose > 0) {
              print($indent . "Command is \"$command\"\n");
            }
                                                               # execute command
            execute_command($configuration{'baseDirectory'}, $verbose, %body);
          }
        }
      }
    }
  }
}

xpl_disconnect($xpl_socket, $xpl_id, $xpl_ip, $client_port);


################################################################################
# Documentation (access it with: perldoc <scriptname>)
#
__END__

=head1 NAME

xpl-actions.pl - Launches scripts based on xPL messages

=head1 SYNOPSIS

xpl-actions.pl [options]

=head1 DESCRIPTION

This xPL client receives xPL messages to launch commands
found in a specific directory.

The C<action.basic> command messages contain following items:

=over 8

=item B<command>

The name of the command to be launched.

=item param=value

parameter/values pairs.

=back

=head1 OPTIONS

=over 8

=item B<-h>

Display a help message.

=item B<-v>

Be verbose.

=item B<-p port>

Specify the base port from which the client searches for a free port.
If not specified, the client will take a default value.

=item B<-n id>

Specify the instance id (name).
The id is limited to 12 characters.
If not specified, it is constructed from the host name and the serial port controller name.

=item B<-t mins>

Specify the number of minutes between two heartbeat messages.

=item B<-w secs>

Specify the number of seconds before sending the first heartbeat.
This allows to start the client after the hub,
thus eliminating an prospective startup delay of one heartbeat interval.

=item B<-d str>

Specify the directory in which the commands are found.

=back

=head1 TEST

Make sure you have the corresponding scripts in the actions directory.

Start the actions launcher:
C<xpl-actions.pl -v -n computer_name>.

You should see messages to the C<xpl-actions.pl>
each time a telegram is sent.

Launch a simple command:
C<xpl-send.pl -v -c actions.basic command=play.bash>.
Launch a command with a short parameter:
C<xpl-send.pl -v -c actions.basic command=play.bash f=pop.wav>.
Launch a command with full parameters:
C<xpl-send.pl -v -c actions.basic command=openJalousie.bash room=study shutter=window pulses=5>.

=head1 AUTHOR

Francois Corthay, DSPC

=head1 VERSION

2.0, 2012

=cut
